README.md
# Carousel component for react-native

## Installation
```bash
yarn add @rolandhordos/react-native-carousel
```

## Usage

### Properties

| Prop | Description | Default |
|---|---|---|
|**`hideIndicators`**|Set to true to hide the indicators |*false*|
|**`indicatorColor `**| Active indicator color |*#FFFFFF*|
|**`indicatorSize `**| Indicator bullet size |*20*|
|**`indicatorSpace `**| Spacing between each indicator |*15*|
|**`inactiveIndicatorColor `**| Inactive indicator color |*#999999*|
|**`indicatorAtBottom `**| Set to false to show the indicators at the top |*true*|
|**`indicatorOffset `**| Indicator relative position from top or bottom |*250*|
|**`onPageChange `**| Called when the active page changes |*callback / method*|
|**`inactiveIndicatorText`**| Inactive indicator content ( You can customize to use any Unicode character ) |*•*|
|**`indicatorText`**| Active indicator content ( You can customize to use any Unicode character ) |*•*|
|**`animate`**| Enable carousel autoplay |*true*|
|**`delay`**| Set Animation delay between slides |*1000*|
|**`loop`**| Allow infinite looped animation. Depends on Prop **`animate`** set to *true*. |*true*|

### Example

```javascript
import React, { Component } from 'react'
import { View, Text, StyleSheet, AppRegistry } from 'react-native'
import Carousel from '@rolandhordos/react-native-carousel'

export default class App extends Component {
  render() {
    return (
      <Carousel animate={false}>
        <View style={styles.slide}>
          <Text>Page 1</Text>
        </View>
        <View style={styles.slide}>
          <Text>Page 2</Text>
        </View>
        <View style={styles.slide}>
          <Text>Page 3</Text>
        </View>
      </Carousel>
    )
  }
}

const styles = StyleSheet.create({
  slide: {
    width: 375,
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'transparent',
  }
})
```

## Project Norms

- Test Supported with Jest
- ES6+
- No Mixins
- Good Mobile Citizen
  - high frequency timers enabled only when required

## Roadmap

### 0.11.5
- Jest Snapshot testing
- Initial pager testing

### 0.11.6
- Beginnings of animation testing
- ScrollViewMock enhancements


CHANGELOG.md
### 0.11.4
- Release 0.11.4 fixed bound method for momentum scroll event
- Fixed and simplified the inline sample code for current React Native

### 0.11.2-3
- Simple Carousel static render test passing
- ES6+ used throughout with current React-Native conventions
- React 16 deprecation removal
- react-timer-mixin replaced with 'react-native-timer'
- package restructuring for transpiled release npm
- 'prepack' action added, for use before 'pack'

### 0.11.1 alpha
- Work in Progress - packaging for client testing
- component evaluation for timer to replace mixin
- simple render test passing

### 0.11.0 alpha
- fork from (@nick)
- DRY CarouselPager Refactor (@echoplans)
- initial scoped NPM @rolandhordos/react-native-carousel

### 0.10.0
- Fix horizontal scroll in RN v0.28 (@sghiassy, @dmccer)

### 0.9.1
- onPageChange never called on indicator press on Android (@lideen)

### 0.9.0
- Fix carousel in iOS with react-native 0.28 (@francisco-sanchez-molina)

### 0.8.0
- Custom indicator content (@shidhincr)

### 0.7.0
- Call createClass on React instead of React Native (@scottluptowski)

### 0.6.1
- Hide inactive indicators (@scottluptowski)

### 0.6.0
- Compute width on the fly (@mikelambert)
- Android support (@mikelambert, @kushal)

### 0.5.0
- Fix for pressing indicator with custom width (@juankiz)
- Fix scrollWithoutAnimationTo deprecation warning (@juankiz)

### 0.4.6
- Fix depreciated parameter for scrollTo (@ronaldx)

### 0.4.5
- Fix scrolling to top when pressing iOS status bar (@juankiz)

### 0.4.4
- Added indicatorOffset property (@regeda)

### 0.4.3
- Fix for page loop animation

### 0.4.2
- Fix for Dimension import

### 0.4.1
- Remove duplicate width calculation when initialPage is defined (@ptmt)

### 0.4.0
- Add autoplay and looping options (@PraiseGeek)

### 0.3.5
- Add `indicatorSpace` prop (@alinz)

### 0.3.3
- Add `indicatorSize` prop (@daviferreira)

### 0.3.2
- Add `hideIndicators` prop (@daviferreira)

### 0.3.1
- Add `onPageChange` prop (@unknownexception)

### 0.3.0
- Bump react-native dependency

### 0.2.0
- Add `indicatorAtBottom` prop (@imaffett)
- Added new styles for the  indicator positions (@imaffett)
- Clicking the indicators will switch the card/page it is on. (@imaffett)
- Set the background of the indicator to transparent. (@imaffett)
- Set bounces=false on the ScrollView used for the Carousel (@imaffett)

### 0.1.0
- Initial component


__tests__/Carousel.js
import React from 'react'
import { View, Text } from 'react-native'
import Carousel from '../src/Carousel'
import renderer from 'react-test-renderer'

describe('Simple Carousel', () => {

  const carousel = renderer.create(
    <Carousel width={250} animate={false}>
      <View><Text>Slide 1</Text></View>
      <View><Text>Slide 2</Text></View>
      <View><Text>Slide 3</Text></View>
    </Carousel>
  )

  it('renders a simple configuration', () => {
    expect(carousel).toBeDefined()

    const structure = carousel.toJSON()
    expect(structure).toBeDefined()
  })
})


src/CarouselPager.js
import React, { Component } from 'react'
import { ScrollView } from 'react-native'

export default class CarouselPager extends Component {

  scrollToPage = (page, animated) => {
    if (typeof animated === 'undefined') {
      animated = true;
    }
    this.refs.scrollView.scrollTo({x: page * this.props.width, y: 0, animated: animated});
  }

  _onMomentumScrollEnd = (e) => {
    var activePage = e.nativeEvent.contentOffset.x / this.props.width;
    this.props.onEnd(activePage);
  }

  render() {
    return <ScrollView ref="scrollView"
      contentContainerStyle={this.props.contentContainerStyle}
      automaticallyAdjustContentInsets={false}
      horizontal={true}
      pagingEnabled={true}
      showsHorizontalScrollIndicator={false}
      bounces={false}
      onScrollBeginDrag={this.props.onBegin}
      onMomentumScrollEnd={this._onMomentumScrollEnd}
      scrollsToTop={false}
    >
      {this.props.children}
    </ScrollView>;
  }
}


Carousel.js
import React, { Component } from 'react'
import { View, Text, Dimensions, StyleSheet } from 'react-native'
import CarouselPager from './CarouselPager'
import timer from 'react-native-timer'  // exported as an instance

export default class Carousel extends Component {

  static defaultProps = {
      hideIndicators: false,
      indicatorColor: '#000000',
      indicatorSize: 50,
      inactiveIndicatorColor: '#999999',
      indicatorAtBottom: true,
      indicatorOffset: 250,
      indicatorText: '•',
      inactiveIndicatorText: '•',
      width: null,
      initialPage: 0,
      indicatorSpace: 25,
      animate: true,
      delay: 1000,
      loop: true,
  }

  state = {
    activePage: this.props.initialPage > 0 ? this.props.initialPage : 0
  }

  static defaultTimerName = 'animation'

  getWidth() {
    if (this.props.width !== null) {
      return this.props.width;
    } else {
      return Dimensions.get('window').width;
    }
  }

  componentDidMount() {
    if (this.props.initialPage > 0) {
      this.refs.pager.scrollToPage(this.props.initialPage, false);
    }

    if (this.props.animate && this.props.children){
        this._setUpTimer();
    }
  }

  indicatorPressed(activePage) {
    this.setState({activePage});
    this.refs.pager.scrollToPage(activePage);
  }

  renderPageIndicator() {
    if (this.props.hideIndicators === true) {
      return null;
    }

    var indicators = [],
        indicatorStyle = this.props.indicatorAtBottom ? { bottom: this.props.indicatorOffset } : { top: this.props.indicatorOffset },
        style, position;

    position = {
      width: this.props.children.length * this.props.indicatorSpace,
    };
    position.left = (this.getWidth() - position.width) / 2;

    for (var i = 0, l = this.props.children.length; i < l; i++) {
      if (typeof this.props.children[i] === "undefined") {
        continue;
      }

      style = i === this.state.activePage ? { color: this.props.indicatorColor } : { color: this.props.inactiveIndicatorColor };
      indicators.push(
         <Text
            style={[style, { fontSize: this.props.indicatorSize }]}
            key={i}
            onPress={this.indicatorPressed.bind(this,i)}
          >
             { i === this.state.activePage  ? this.props.indicatorText : this.props.inactiveIndicatorText }
          </Text>
      );
    }

    if (indicators.length === 1) {
      return null;
    }

    return (
      <View style={[styles.pageIndicator, position, indicatorStyle]}>
        {indicators}
      </View>
    );
  }

  _setUpTimer() {
     if (this.props.children.length > 1) {
       timer.setTimeout(Carousel.defaultTimerName, this._animateNextPage, this.props.delay);
     }
  }

  _animateNextPage = () => {
     var activePage = 0
     if (this.state.activePage < this.props.children.length - 1) {
         activePage = this.state.activePage + 1;
     } else if (!this.props.loop) {
         return;
     }

     this.indicatorPressed(activePage);
     this._setUpTimer();
  }

  _onAnimationBegin = () => {
     timer.clearTimeout(Carousel.defaultTimerName)
  }

  _onAnimationEnd = (activePage) => {
    this.setState({activePage});
    if (this.props.onPageChange) {
      this.props.onPageChange(activePage);
    }
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <CarouselPager
          ref="pager"
          width={this.getWidth()}
          contentContainerStyle={styles.container}
          onBegin={this._onAnimationBeginPage}
          onEnd={this._onAnimationEnd}
        >
          {this.props.children}
        </CarouselPager>
        {this.renderPageIndicator()}
      </View>
    )
  }

}

const styles = StyleSheet.create({
  pageIndicator: {
    position: 'absolute',
    flexDirection: 'row',
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
    alignSelf: 'center',
    backgroundColor:'transparent',
  }
})
